src/parser.js

```diff
@@ -0,0 +1,104 @@
+'lang sweet.js';
+// @flow
+import { Either, Left } from './either';
+import { Maybe, Just, Nothing } from './maybe';
+
+type Mark = {};
+
+interface Context<T> {
+  next(): {
+    done: boolean;
+    value: T
+  };
+
+  mark(): Mark;
+
+  reset(Mark): void;
+}
+
+export class ImmutableContext<T> {
+  ctx: Context<T>;
+  value: Maybe<T>;
+
+  constructor(ctx: Context<T>) {
+    this.ctx = ctx;
+    let mark = ctx.mark();
+    let v = ctx.next();
+    ctx.reset(mark);
+    this.value = v.done ? new Nothing() : Maybe.of(v.value);
+  }
+
+  head() {
+    return this.value;
+  }
+
+  rest(): ImmutableContext<T> {
+    this.ctx.next();
+    return new ImmutableContext(this.ctx); 
+  }
+}
+
+export default class Parser<A, C> {
+  runner: ImmutableContext<C> => Either<string, [A, ImmutableContext<C>]>;
+
+  constructor(runner: ImmutableContext<C> => Either<string, [A, ImmutableContext<C>]>) {
+    this.runner = runner;
+  }
+  
+  run(ctx: ImmutableContext<*>): Either<string, [A, ImmutableContext<*>]> {
+    return this.runner(ctx);
+  }
+
+  static of(a: A): Parser<A, *> {
+    return new Parser(ctx => Either.of([a, ctx]));
+  }
+
+  static zero(msg: string = ''): Parser<*, *> {
+    return new Parser(ctx => { return new Left(msg); });
+  }
+
+  static item(): Parser<*, *> {
+    return new Parser(ctx => {
+      let i = ctx.head();
+      if (i instanceof Just) {
+        return Either.of([i.value, ctx.rest()]);
+      } else {
+        return new Left('');
+      }
+    });
+  }
+
+  static sat<T>(f: T => boolean, msg: string = 'Did not satisfy predicate'): Parser<T, T> {
+    return Parser
+      .item()
+      .chain(item => f(item) ? Parser.of(item) : Parser.zero(msg));
+  }
+
+  static seq<A, B, V>(f: (A, B) => V, a: Parser<A, *>, b: Parser<B, *>): Parser<V, *> {
```

Comment:
This is `lift2` (if `f` is curried). You could either define it directly or make `Parser` an Applicative Functor and  implement `ap`.